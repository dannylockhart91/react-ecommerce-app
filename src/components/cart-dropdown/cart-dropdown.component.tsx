import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { createStructuredSelector } from 'reselect';

import { AppState } from '../../store/root.reducer';
import { getCartItems } from '../../store/cart/cart.selectors';
import { toggleShowCart } from '../../store/cart/cart-actions';
import { Item } from '../../store/cart/cart.types';

import CartItem from '../cart-item/cart-item.component';
import CustomButton from '../custom-button/custom-button.component';

import { CartDropdownProps, cartDropdownSelectors } from './cart-dropdown.types';
import { CartDropdownContainer } from './cart-dropdown-styles';

export const CartDropdown: React.FC<CartDropdownProps> = ({ cartItems, dispatch, history }) => (
    <CartDropdownContainer className='cart-dropdown'>
        <div className='cart-items' id="cart-items">
            {cartItems.length ? (
                cartItems.map((item: Item) => <CartItem key={item.id} item={item}/>)
            ) : (
                <span className='empty-message'>Your cart is empty</span>
            )}
        </div>
        <CustomButton
            handleClick={() => {
                history.push('/checkout');
                dispatch(toggleShowCart());
            }}>
            GO TO CHECKOUT
        </CustomButton>
    </CartDropdownContainer>
);

const mapStateToProps = createStructuredSelector<AppState, cartDropdownSelectors>({
    cartItems: getCartItems
});

export default withRouter(connect(mapStateToProps)(CartDropdown));
