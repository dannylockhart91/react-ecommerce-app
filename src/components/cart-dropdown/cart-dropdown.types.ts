import { Item } from "../../store/cart/cart.types";
import { RouteComponentProps } from "react-router";
import { DispatchProp } from "react-redux";

export interface CartDropdownProps extends RouteComponentProps, DispatchProp {
	cartItems: Item[];
}
export interface cartDropdownSelectors {
	cartItems: Item[];
}