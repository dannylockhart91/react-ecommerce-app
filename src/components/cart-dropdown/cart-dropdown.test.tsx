import React from 'react';
import { shallow } from 'enzyme';

import { CartDropdown } from './cart-dropdown.component';

describe('CartDropdown component', () => {
	let wrapper;
	let mockProps = {};

	beforeEach(() => {
		mockProps = {
			cartItems: [
				{
					id: 1,
					name: 'test',
					price: 8,
					imageUrl: 'test.test.com',
					quantity: 1
				}
			],
			cartItemCount: 1
		};
	});

	it('render CartDropdown component', () => {
		mockProps = {
			cartItems: [],
			cartItemCount: 2
		};
		wrapper = shallow(
			//@ts-ignore
			<CartDropdown {...mockProps} />
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('should display empty cart message if there are no items to display', () => {
		mockProps = {
			cartItems: [],
			cartItemCount: 0
		};
		//@ts-ignore
		wrapper = shallow(<CartDropdown {...mockProps} />);
		expect.assertions(1);
		expect(
			wrapper
				.find('[id="cart-items"]')
				.html()
				.includes('Your cart is empty')
		).toEqual(true);
	});

	it('should display item and count when there are items to show', () => {
		//@ts-ignore
		wrapper = shallow(<CartDropdown {...mockProps} />);
	});
});
