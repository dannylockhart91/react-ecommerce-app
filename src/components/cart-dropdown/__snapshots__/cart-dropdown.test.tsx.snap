// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component render CartDropdown component 1`] = `
<styled.div
  className="cart-dropdown"
>
  <div
    className="cart-items"
    id="cart-items"
  >
    <span
      className="empty-message"
    >
      Your cart is empty
    </span>
  </div>
  <CustomButton
    handleClick={[Function]}
  >
    GO TO CHECKOUT
  </CustomButton>
</styled.div>
`;
